bin/pr_changed_files

```diff
@@ -0,0 +1,129 @@
+#   pr_changed_files --only-in <file-patterns...>     # Check if changes are limited to given patterns
+#
+# Examples:
+#   # Check if any files changed
+#   pr_changed_files
+#
+#   # Check if only documentation files changed
+#   pr_changed_files --only-in "*.md" "docs/**"
+#
+#   # Check if any Swift files changed
+#   pr_changed_files --includes "*.swift"
+#
+# Returns:
+#   Prints "true" if the condition is met, "false" otherwise
+#   Always exits with code 0 to allow easy variable assignment
+
+if [[ ! "${BUILDKITE_PULL_REQUEST:-invalid}" =~ ^[0-9]+$ ]]; then
+    echo "Error: this tool can only be called from a Buildkite PR job" >&2
+    exit 1
+fi
+
+mode=""
+patterns=()
+
+# Define error message for mutually exclusive options
+EXCLUSIVE_OPTIONS_ERROR="Error: either specify --only-in or --includes; cannot specify both"
+
+while [[ "$#" -gt 0 ]]; do
+    case $1 in
+        --only-in)
+            if [[ -n "$mode" ]]; then
+                echo "$EXCLUSIVE_OPTIONS_ERROR" >&2
+                exit 1
+            fi
+            mode="only-in"
+            shift
+            while [[ "$#" -gt 0 && "$1" != "--"* ]]; do
+                patterns+=("$1")
+                shift
+            done
+            ;;
+        --includes)
+            if [[ -n "$mode" ]]; then
+                echo "$EXCLUSIVE_OPTIONS_ERROR" >&2
+                exit 1
+            fi
+            mode="includes"
+            shift
+            while [[ "$#" -gt 0 && "$1" != "--"* ]]; do
+                patterns+=("$1")
+                shift
+            done
+            ;;
+        --*) 
+            echo "Error: unknown option $1" >&2
+            exit 1
+            ;;
+        *)
+            echo "Error: unexpected argument $1" >&2
+            exit 1
+            ;;
+    esac
+done
+
+# Get list of changed files
+changed_files=$(git --no-pager diff --name-only --merge-base "$BUILDKITE_PULL_REQUEST_BASE_BRANCH" HEAD | sort)
+
+if [[ -z "$mode" ]]; then
+    # No arguments = any change
+    if [[ -n "$changed_files" ]]; then
+        echo "true"
+    else
+        echo "false"
+    fi
+    exit 0
+fi
+
+if [[ ${#patterns[@]} -eq 0 ]]; then
+    echo "Error: must specify at least one file pattern" >&2
+    exit 1
+fi
+
+# No files changed
+if [[ -z "$changed_files" ]]; then
+    if [[ "$mode" == "only-in" ]]; then
+        echo "true"
+    else
+        echo "false"
+    fi
+    exit 0
+fi
+
+if [[ "$mode" == "only-in" ]]; then
+    # Check if all changed files match at least one pattern
+    for file in $changed_files; do
+        matches_any=false
+        for pattern in "${patterns[@]}"; do
+            # shellcheck disable=SC2053
+            if [[ "$file" == $pattern ]]; then # glob match with pattern
+                matches_any=true
+                break
+            fi
+        done
+        if [[ "$matches_any" == "false" ]]; then
+            echo "false"
+            exit 0
+        fi
+    done
+    echo "true"
```

Comment:
One way to maybe make this more readable would be to extract the test of if `$file` matches at least one of the `$patterns`… in a dedicated function, like `file_matches_any_pattern` that would return `0` if match found and `1`  if not.

Then you'd be able to use it:
 - Here inside your `for file in "$changed_files"` loop for `only-in` to just say `if ! file_matches_any_pattern "$file" "$patterns[@]"; then echo "false"; exit 0; fi`, and still with the `echo "true"` outside the file loop for if all files passed the test
 - In the `includes` mode by using `if file_matches_any_pattern; then echo "false"; exit 0; fi", still with the `echo "false"` outside the file loop for if none of the files passed the test